### ISSUE TITLE
Loss drops to 0 every 50 epochs on TOX21

### ISSUE THREAD

**Issue by guillaumeRenton** [deepchem/deepchem#1944]
Hello,

I'm currently using deepChem 2.3.0.

I have troubles when attempting to train on the TOX21 dataset. I tried multiple models (currently GraphConv, Weaver and MPNN), and for each one, my loss drop to 0 every 50 epochs.

Any idea what could cause this ? 

Here is a little code to reproduce the problem with GraphConvModel.

```python
import numpy as np
import deepchem as dc
from deepchem.molnet import load_tox21
from deepchem.models.graph_models import GraphConvModel
import matplotlib.pyplot as plt

model_dir = "/tmp/graph_conv"

tox21_tasks, tox21_datasets, transformers = load_tox21(featurizer='GraphConv')
train_dataset, valid_dataset, test_dataset = tox21_datasets

metric = dc.metrics.Metric(
    dc.metrics.roc_auc_score, np.mean, mode="classification")

batch_size = 50

model = GraphConvModel(
    len(tox21_tasks), batch_size=batch_size, mode='classification')

losses = []

for i in range(105):
    loss = model.fit(train_dataset, nb_epoch=1)
    losses.append(loss)

print(losses)
plt.plot(losses)
plt.show()
```

**Comment by rbharath** [comment]
Thanks for reporting this! I'll look at this shortly and report back :)

**Comment by rbharath** [comment]
I looked into this and I'm getting an even worse error on HEAD:

```
losses == [None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None]
```

That is, losses aren't even being returned for some reason on HEAD. I've marked this as a bug.

**Comment by rbharath** [comment]
I made a mistake in my previous analysis. I ran this more carefully and I'm getting the following losses:

```
losses == [0.9411619120630724, 0.8790051032756937, 0.8168775185771372, 0.7157551050186157, 0.715576171875, 0.6869997591585726, 0.5502074956893921, 0.6697917580604553, 0.6221965101898693, 0.6009211646185981, 0.6114039170114618, 0.5842803319295248, 0.5567047812721946, 0.4655408064524333, 0.5476418086460658, 0.5234002470970154, 0.5100434826266381, 0.46794197776100854, 0.48683226342294733, 0.46727547645568845, 0.46957556406656903, 0.4511328245464124, 0.4482121538760057, 0.4263594150543213, 0.41964977264404296, 0.4313131968180339, 0.40900641751576616, 0.38885780175526935, 0.40570984817132716, 0.38564371381487167, 0.3867200600980508, 0.3961425168173654, 0.3810104570890728, 0.37083543733108876, 0.34530436197916664, 0.36292472752657806, 0.3404340090816968, 0.22491499781608582, 0.32054405827676097, 0.3327127774556478, 0.3355512726172972, 0.31377191013760036, 0.32735071790979264, 0.3113608109323602, 0.2878539800643921, 0.32444042318007527, 0.3318012479751829, 0.3000284899836001, 0.2947992143176851, 0.27717884063720705, 0.28183685978756673, 0.3111102879047394, 0.30921938612654404, 0.2802161592425722, 0.26781156439530224, 0.28038014968236286, 0.26669212557234856, 0.25320334550811024, 0.25236131928183814, 0.25508620738983157, 0.2647107165792714, 0.2531622672567562, 0.2508337939227069, 0.24497011729649135, 0.23823924345128678, 0.2552474226270403, 0.25156524569489236, 0.24123483233981663, 0.0893775001168251, 0.22594923973083497, 0.22102362422619837, 0.23621901598843661, 0.2706871032714844, 0.22959740265556003, 0.23914077758789062, 0.16377152502536774, 0.21078540339614404, 0.2276399673954133, 0.22486751682155734, 0.20937697887420653, 0.1959281065026108, 0.1993839190556453, 0.2039637225014823, 0.1961843172709147, 0.19295685107891375, 0.19193825823195437, 0.20362219603165335, 0.17934065598707932, 0.20033452540268132, 0.20918686389923097, 0.19115443107409355, 0.18895240391001983, 0.20014462028582072, 0.20122489562401405, 0.1899304996837269, 0.17897605895996094, 0.1944283888890193, 0.16937009493509927, 0.173996347776601, **0.0,** 0.18096542358398438, 0.17388246799337453, 0.16662793872000156, 0.16150125861167908, 0.15584423277113174]
```
It's hard to see but `losses[99] == 0` (I've bolded it). I'm also seeing user reports of getting zeros every 5 epochs even. 

This bug is confusing since I looked through the code and I'm not seeing any obvious reason why the loss could suddenly drop to 0. It's also not consistent. @guillaumeRenton is reporting every 50 epochs. I'm seeing a 0 on the 100th epoch and other users report every 5 epochs.

@peastman do you have any ideas here?

**Comment by guillaumeRenton** [comment]
Could the periodicity depends on the dataset used ? Or is everyone getting the problem only on TOX21 ?

**Comment by rbharath** [comment]
It seems to happen for other datasets too. @chstem reported seeing it every 5 epochs on his dataset (not Tox21)

**Comment by rbharath** [comment]
Ah, I've figured it out! It's due to these lines:

https://github.com/deepchem/deepchem/blob/master/deepchem/models/keras_model.py#L393. For ease, reproduced:

```
if should_log:
  avg_loss = float(avg_loss) / averaged_batches
  logger.info(
      'Ending global_step %d: Average loss %g' % (current_step, avg_loss))
  avg_loss = 0.0
```

Essentially the average loss is reset to 0 after output is logged. Depending on what you've set as the log frequency, this mean you see loss 0.0 that often. This is harmless for actual model health, but I think we could add a workaround so we don't have distressing drops in the loss curves

**Comment by chstem** [comment]
I have two suggestions about this `log_frequency`:

- Setting `log_frequency = 0` should disable logging. Currently it will cause a zero division warning and return of zero loss. This could also serve as a workaround, but it is of course not a fix.
- It seems more intuitive to set this parameter in `model.fit()`, not in `model.__init__()`.

**Comment by rbharath** [comment]
Yep +1 to both suggestions I think

**Comment by rbharath** [comment]
I'm making this issue open for contribution since we have a pretty good idea of what's happening here. (If no one else fixes this, I'm glad to this myself, but there are some more critical bugs that are higher up on my workstream for now)

**Comment by rbharath** [comment]
This issue should have been fixed by #2060. Just upgrade to DeepChem's latest version with `pip install --upgrade --pre deepchem` and you should be set. I'll keep this issue open in case there are still any issues

**Comment by chstem** [comment]
I just tested this, but I still have the issue.

Here are some numbers:
If I use 80000 training samples  and a batch_size of 100 I get 0 for all loses.
If I reduce the training samples to 8000, only every 5th loss is 0. If I additionally set batch_size to 32, every second loss is zero. With batch_size=8, again all loses are zero.

Deepchem version according to pip: 2.4.0rc1.dev2020080308371

**Comment by Draghohmmar** [comment]
Used tp gey the same issue every 10 epochs, using standard GraphConv set-up:
![image](https://user-images.githubusercontent.com/47124775/89161410-7d150280-d56a-11ea-8e34-0a186e339866.png)

but running "pip install --upgrade --pre deepchem", fixed it for me

**Comment by rbharath** [comment]
@chstem Hmm, that's disconcerting. I don't think I've seen 0 for all losses. Would it be possible for you to test on a standard dataset like Tox21? I'm not currently seeing the 0s on any of my test datasets, but it's very possible I'm missing something. I'll do some more thorough tests to see if I can reproduce on my end

**Comment by chstem** [comment]
@rbharath I could reproce it with Tox21 quite easily. I just construct some dataset of length 8000 (to match my numbers above):

```
import deepchem as dc
from deepchem.models.graph_models import GraphConvModel

tox21_tasks, tox21_datasets, transformers = dc.molnet.load_tox21(featurizer='GraphConv', reload=False)
train_dataset, valid_dataset, test_dataset = tox21_datasets
n_tasks = len(tox21_tasks)

ds2 = train_dataset.select(range(2000))
ds6 = train_dataset.select(range(6000))
ds8 = train_dataset.merge((ds2, ds6))


model = GraphConvModel(n_tasks, batch_size=8, mode='classification')
num_epochs = 20
losses = []
for i in range(num_epochs):
    loss = model.fit(ds8, nb_epoch=1)
    print("Epoch %d loss: %f" % (i, loss))
    losses.append(loss)
```

This code prints all zero losses for me. Increasing `batch_size` will reduce the frequency of reported zero losses.

**Comment by rbharath** [comment]
@chstem I tried running this code on the latest DeepChem and I'm getting no zero losses. Would you mind double checking that you're using the latest DeepChem install in a fresh environment? It might be an upgrade issue where an older DeepChem is still being used for some reason. You can check this easily by seeing if the new `all_losses` feature for loss curves is enabled (this was added in the same PR as the loss function fix). Here's a slight modification of your code snippet:

```
import deepchem as dc
from deepchem.models.graph_models import GraphConvModel

tox21_tasks, tox21_datasets, transformers = dc.molnet.load_tox21(featurizer='GraphConv', reload=False)
train_dataset, valid_dataset, test_dataset = tox21_datasets
n_tasks = len(tox21_tasks)

ds2 = train_dataset.select(range(2000))
ds6 = train_dataset.select(range(6000))
ds8 = train_dataset.merge((ds2, ds6))


model = GraphConvModel(n_tasks, batch_size=8, mode='classification')
num_epochs = 20
losses = []
all_losses = []
for i in range(num_epochs):
    loss = model.fit(ds8, nb_epoch=1, all_losses=all_losses)
    print("Epoch %d loss: %f" % (i, loss))
    losses.append(loss)
```

`all_losses` contains the losses for every reported loss in the logs and provides a more fine grained view of the loss curve.

**Comment by chstem** [comment]
You were right, somehow I didn't have the deepchem version pip was claiming I have.
Now I can confirm the issues is fixed.

**Comment by rbharath** [comment]
Awesome :). Going to go ahead and close off this issue. Will reopen in case there are any more reports of this still being broken
